openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp

```diff
@@ -1872,6 +1873,38 @@ struct AMDGPUDeviceTy : public GenericDeviceTy, AMDGenericDeviceTy {
     else
       return Plugin::error("Unexpected AMDGPU wavefront %d", WavefrontSize);
 
+    // To determine the correct scratch memory size per thread, we need to check
+    // the device architecure generation. According to AOT_OFFLOADARCHS we may
+    // assume that AMDGPU offload archs are prefixed with "gfx" and suffixed
+    // with a two char arch specialization. In-between is the 1-2 char
+    // generation number we want to extract.
+    std::string CUKind{ComputeUnitKind};
```

Comment:
This code looks really complicated and should be using `StringRef`. Are we just trying to get the major version? That should be pretty trivial.
```c++
StringRef Arch(ComputeUnitKind)
unsigned Version = 0u;
if (llvm::to_integer(Arch.drop_front(sizeof("gfx") - 1), Version))
  return Plugin::error("Invalid GFX architecture string");
unsigned Major = Version / 100;
```

Reply:
Yes we only want the major version, so 9 for `gfx90a` or 10 for `gfx1030`.
It is pretty complicated for what it actually does. With the C++ tools I was aware of, I wanted to make sure I do not produce garbage.
Thank you very much!

Reply:
Sorry, I forgot a `!` on that condition, the string conversion returns false on failure.

Reply:
Thanks for the correction!
Also, e.g. "gfx90a" will fail, so I'd go for a slice:
```C++
    StringRef Arch(ComputeUnitKind);
    unsigned GfxGen = 0u;
    if (!llvm::to_integer(Arch.slice(sizeof("gfx") - 1, Arch.size() - 2),
                          GfxGen))
```
In any way, this is significantly easier to read.